feat(core): execution strategy (#758)
MarathonLabs/marathon: docs/docs/intro/configure.md

</TabItem>
</Tabs>

### Execution strategy
When executing tests with retries there are multiple trade-offs to be made. Two execution strategies are supported: any success or all success.

> By default, any success strategy is used with fast execution i.e. if one of the test retries succeeds then the test is considered successfully

Reviewer: Worth changing all instances of "any success" & "all success" to `ANY_SUCCESS` & `ALL_SUCCESS` for easier reading?

executed and all non-started retries are removed.

#### Any success

> Test passes if any of executions is passing. This mode works only if there is no complex sharding strategy applied. This is the default.

Reviewer: ```suggestion
Test passes if any of its executions are passing. This mode works only if there is no complex sharding strategy applied. This is the default.
```

Author: done

<Tabs>
<TabItem value="YAML" label="Marathonfile">

```yaml
executionStrategy:
  mode: ANY_SUCCESS
```

</TabItem>
<TabItem value="kts" label="Kotlin DSL">

```kotlin
marathon {
  executionStrategy = ExecutionStrategyConfiguration(ExecutionMode.ANY_SUCCESS)
}
```

</TabItem>
<TabItem value="groovy" label="Groovy DSL">

```groovy
marathon {
  executionStrategy = ExecutionStrategyConfiguration(ExecutionMode.ANY_SUCCESS)
}
```

</TabItem>
</Tabs>

:::info

Complex sharding with any success mode doesn't make sense when user asks for N number of tests to run explicitly, and we pass on the first one.

:::

#### All success

> Test passes if and only if all the executions are passing. This mode works only if there are no retries, i.e. no complex flakiness strategy, no retry strategy.

Reviewer: ```suggestion
Test passes if and only if all its executions are passing. This mode works only if there are no retries, i.e. no complex flakiness strategy, no retry strategy.
```

Author: done

<Tabs>
<TabItem value="YAML" label="Marathonfile">

```yaml
executionStrategy:
  mode: ALL_SUCCESS
```

</TabItem>
<TabItem value="kts" label="Kotlin DSL">

```kotlin
marathon {
  executionStrategy = ExecutionStrategyConfiguration(ExecutionMode.ALL_SUCCESS)
}
```

</TabItem>
<TabItem value="groovy" label="Groovy DSL">

```groovy
marathon {
  executionStrategy = ExecutionStrategyConfiguration(ExecutionMode.ALL_SUCCESS)
}
```

</TabItem>
</Tabs>

:::info

> Adding retries with retry/flakiness strategies means users wants to trade-off cost for reliability, i.e. add more retries and pass if one 

Reviewer: ```suggestion
Adding retries with retry/flakiness strategies means users want to trade off cost for reliability, i.e. add more retries and pass if one 
```

Also, Is execution time the only "cost" being referred to here and elsewhere on this page? If so, perhaps we can be more specific.

Author: Cost is how much one pays with $ for the test runs

Author: changed the wording as suggested

of test retries passes, so retries only make sense for any success mode. 

> When we use all success mode it means user wants to verify each test with a number of tries (they are not retries per se) and pass only if 

Reviewer: ```suggestion
When we use all success mode it means the user wants to verify each test with a number of tries (they are not retries per se) and pass only if 
```

Author: done

all of them succeed. This is the case when fixing a flaky test or adding a new test, and we want to have a signal that the test is fixed/not flaky.

:::

#### Fast execution mode

> When test reaches a state when a decision about it's state can be made marathon can remove additional in-progress retries.

Reviewer: ```suggestion
When the test reaches a state where a decision about its state can be made, marathon can remove additional in-progress retries.
```

Author: done

This decision point is different depending on the execution mode used. Let's walk through two examples.

> Assume any success strategy is used and 100 retries are scheduled for a test A via flakiness strategy. Let's say first 3 failed and the 4th attempt succeeded. At 

Reviewer: ```suggestion
Assume the any success strategy is used and 100 retries are scheduled for test A via the flakiness strategy. Let's say the first 3 failed and the 4th attempt succeeded. At 
```

> this point the test should already be considered passed since any success out of all retries leads to the result by definition of any successs

Reviewer: Except for the first use of "any success" on this line, is it worth changing all instances of "any success" & "all success" to `ANY_SUCCESS` & `ALL_SUCCESS` for easier reading?

> this point the test should already be considered passed since any success out of all retries leads to the result by definition of any successs

Reviewer: Except for the first use of "any success" on line 468, is it worth changing all instances of "any success" & "all success" in this doc to `ANY_SUCCESS` & `ALL_SUCCESS` for easier reading?

Author: replaced

> execution strategy. To save cost one can remove additional non-started retries by using fast execution mode (this is the default behavior for

Reviewer: Is "time" the only cost being referred to here and elsewhere on this page? If so, perhaps "time" is more descriptive.

any success strategy). On the other hand one could disable fast execution and get much more accurate statistics about this test by executing
more retries and calculating the probability of passing as a measure of flakiness for test A.

> Assume all success strategy is used and 100 retries are scheduled using sharding strategy. Let's say first 3 passed and the 4th attempt failed.

Reviewer: ```suggestion
Assume the all success strategy is used and 100 retries are scheduled using the sharding strategy. Let's say the first 3 passed and the 4th attempt failed.
```
